darts/explainability/shap_explainer.py

```diff
@@ -0,0 +1,474 @@
+            background_series,
+            background_past_covariates,
+            background_future_covariates,
+            background_nb_samples
+        )
+
+        self.target_dict = {c:i for i, c in enumerate(background_series.columns)}
+
+        if (self.target_dim > 1 or self.model.output_chunk_length > 1) and (not self.multioutput):
+            self.explainers = {}
+            for i in range(self.model.output_chunk_length):
+                self.explainers[i] = {}
+                for j in range(self.target_dim):
+                    self.explainers[i][j] = shap.Explainer(
+                        self.model.model.estimators_[i+j], 
+                        self.background_X
+                    )
+                    # Special case of trees, where we don't want to depend on a background (faster in general)
+                    # Warning: Randomforest is quite slow.
+                    if isinstance(self.explainers[i][j], shap.explainers._tree.Tree):
+                        self.explainers[i][j] = shap.TreeExplainer(
+                            self.model.model.estimators_[i+j])
+        else:
+            self.explainers = shap.Explainer(self.model.model, self.background_X)
+            if isinstance(self.explainers, shap.explainers._tree.Tree):
+                self.explainers = shap.TreeExplainer(self.model.model)
+
+        self.cache_explainers = None
+        self.cache_foreground_series = None
+        self.cache_foreground_past_covariates = None
+        self.cache_foreground_future_covariates = None
+        self.cache_foreground_X = None
+        self.foreground_changed = True
+
+    def shap_values(
+        self,
+        foreground_series: Union[TimeSeries, Sequence[TimeSeries]],
+        foreground_past_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        foreground_future_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        horizon: Optional[int]=None, 
+        target_name: Optional[str]=None
+    ) -> shap._explanation.Explanation:
+        
+        raise_if_not(
+            (len(foreground_series) == len(foreground_past_covariates)) and \
+            (len(foreground_past_covariates) == len(foreground_future_covariates)) and \
+            (len(foreground_future_covariates) == len(foreground_series)),
+            'target, past and future covariates must have the same length.'
+        )
+
+        raise_if_not(
+            all(foreground_series.time_index == foreground_past_covariates.time_index) and \
+            all(foreground_past_covariates.time_index == foreground_future_covariates.time_index) and \
+            all(foreground_future_covariates.time_index == foreground_series.time_index),
+            'The target, past and future covariates must have the same time index.'
+        )
+
+        # We don't recompute if the foreground is the same as the last computation
+        if (self.cache_foreground_series != foreground_series) or \
```

Comment:
What does this cache buy us? It introduces some extra complexity and state in the class, and I'm not sure it's worth the price. I would put this burden on the user (not calling twice the same method with the same inputs is user's responsibility), but maybe I'm missing something.